include/CL/opencl.hpp

```diff
@@ -1641,7 +1645,11 @@ struct GetInfoFunctor0
     Func f_; const Arg0& arg0_;
     cl_int operator ()(
         cl_uint param, size_type size, void* value, size_type* size_ret)
+#if defined(CL_USE_FUNCTION_POINTERS)
+    { return (*f_)(arg0_, param, size, value, size_ret); }
+#else
     { return f_(arg0_, param, size, value, size_ret); }
+#endif
```

Comment:
I'm not quite following why this change is needed, but it'd be really great if we could avoid needing an #ifdef for this.  A couple of options to consider:

1. Will the `CL_USE_FUNCTION_POINTERS` path work in all cases?  This seemed to work for the few cases I've tried and see also: https://godbolt.org/z/fMfTedq8d
2. If not, can we do something with partial template specialization when `Func` is a pointer type?

Reply:
Oh, I'm sure this can be done more elegantly, the reason why this is needed is that when using function pointers to access the OpenCL API f_ becomes a pointer to a function pointer instead of a simple function pointer, and therefore you need to dereference once to get the function call syntax to work.

Reply:
Oh, I see what you mean, sure, explicitly dereferencing the pointer does work all the time =)